File: lambUpdateLastest/registerActivity.py
import json
import boto3
import uuid
from datetime import datetime
from botocore.exceptions import ClientError

# Initialize DynamoDB client
dynamodb = boto3.resource('dynamodb')

def lambda_handler(event, context):
    # CORS headers
    headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Content-Type,Authorization',
        'Access-Control-Allow-Methods': 'POST,OPTIONS'
    }
    
    # Handle preflight OPTIONS request
    if event.get('httpMethod') == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': headers,
            'body': ''
        }
    
    try:
        # Parse request body
        if not event.get('body'):
            return {
                'statusCode': 400,
                'headers': headers,
                'body': json.dumps({'success': False, 'message': 'Missing request body'})
            }
        
        request_body = json.loads(event.get('body'))
        
        # ดึงข้อมูลจาก request
        activity_id = request_body.get('activityId')
        student_id = request_body.get('studentId')
        
        if not activity_id or not student_id:
            return {
                'statusCode': 400,
                'headers': headers,
                'body': json.dumps({'success': False, 'message': 'activityId และ studentId จำเป็นต้องระบุ'})
            }
        
        # เชื่อมต่อ DynamoDB tables
        activities_table = dynamodb.Table('Activities')
        participations_table = dynamodb.Table('ActivityParticipations')
        students_table = dynamodb.Table('Students')
        
        # ตรวจสอบว่ามีกิจกรรมนี้อยู่จริง
        activity_response = activities_table.get_item(Key={'activityId': activity_id})
        if 'Item' not in activity_response:
            return {
                'statusCode': 404,
                'headers': headers,
                'body': json.dumps({'success': False, 'message': 'ไม่พบกิจกรรมที่ระบุ'})
            }
        
        activity = activity_response['Item']
        
        # ตรวจสอบว่ามีนักศึกษานี้อยู่จริง
        student_response = students_table.get_item(Key={'studentId': student_id})
        if 'Item' not in student_response:
            return {
                'statusCode': 404,
                'headers': headers,
                'body': json.dumps({'success': False, 'message': 'ไม่พบข้อมูลนักศึกษา'})
            }
        
        # ตรวจสอบว่าได้ลงทะเบียนกิจกรรมนี้แล้วหรือไม่
        existing_participation = participations_table.scan(
            FilterExpression='activityId = :activityId AND studentId = :studentId',
            ExpressionAttributeValues={
                ':activityId': activity_id,
                ':studentId': student_id
            }
        )
        
        if existing_participation.get('Items'):
            return {
                'statusCode': 400,
                'headers': headers,
                'body': json.dumps({'success': False, 'message': 'คุณได้ลงทะเบียนกิจกรรมนี้แล้ว'})
            }
        
        # ตรวจสอบว่ากิจกรรมยังไม่ผ่านมา
        if not is_future_activity(activity.get('startDateTime')):
            return {
                'statusCode': 400,
                'headers': headers,
                'body': json.dumps({'success': False, 'message': 'ไม่สามารถลงทะเบียนกิจกรรมที่ผ่านมาแล้วได้'})
            }
        
        # สร้าง participation record ใหม่
        participation_id = f"part{str(uuid.uuid4()).replace('-', '')[:8]}"
        current_time = datetime.now().isoformat() + 'Z'
        
        participation_item = {
            'participationId': participation_id,
            'activityId': activity_id,
            'studentId': student_id,
            'registeredAt': current_time,
            'isConfirmed': False,  # ยังไม่ได้ยืนยันการเข้าร่วม
            'surveyCompleted': False,  # ยังไม่ได้ทำแบบประเมิน
            'createdAt': current_time,
            'updatedAt': current_time
        }
        
        # บันทึกลง DynamoDB
        participations_table.put_item(Item=participation_item)
        
        print(f'Successfully registered student {student_id} for activity {activity_id}')
        
        return {
            'statusCode': 200,
            'headers': headers,
            'body': json.dumps({
                'success': True,
                'message': 'ลงทะเบียนเข้าร่วมกิจกรรมสำเร็จ',
                'participationId': participation_id,
                'activityName': activity.get('name', ''),
                'note': 'กรุณาเข้าร่วมกิจกรรมและสแกน QR Code เพื่อยืนยันการเข้าร่วม'
            })
        }
        
    except json.JSONDecodeError:
        return {
            'statusCode': 400,
            'headers': headers,
            'body': json.dumps({'success': False, 'message': 'รูปแบบ JSON ไม่ถูกต้อง'})
        }
    except ClientError as e:
        print('DynamoDB Error:', str(e))
        return {
            'statusCode': 500,
            'headers': headers,
            'body': json.dumps({'success': False, 'message': 'เกิดข้อผิดพลาดในฐานข้อมูล', 'details': str(e)})
        }
    except Exception as e:
        print('Unexpected error:', str(e))
        return {
            'statusCode': 500,
            'headers': headers,
            'body': json.dumps({'success': False, 'message': 'เกิดข้อผิดพลาดที่ไม่คาดคิด', 'details': str(e)})
        }

def is_future_activity(start_date_time):
    """ตรวจสอบว่ากิจกรรมยังไม่ผ่านมาหรือไม่"""
    if not start_date_time:
        return True
    
    try:
        # แปลง string เป็น datetime object
        activity_date = datetime.fromisoformat(start_date_time.replace('Z', '+00:00'))
        current_date = datetime.now(activity_date.tzinfo)
        
        # ถ้ากิจกรรมยังไม่ผ่านมา
        return activity_date > current_date
    except:
        # ถ้าไม่สามารถแปลงวันที่ได้ ให้อนุญาต
        return True

File: lambUpdateLastest/registerActivity.mjs
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, PutCommand, QueryCommand } from "@aws-sdk/lib-dynamodb";
import { randomUUID } from "crypto"; // ใช้สร้าง ID ไม่ซ้ำ

const client = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(client);

const TABLE_NAME = "ActivityParticipations";

export const handler = async (event) => {
  // ตั้งค่า Headers สำหรับ CORS (เพื่อให้หน้าเว็บเรียกใช้งานได้)
  const headers = {
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Headers": "Content-Type,Authorization",
    "Access-Control-Allow-Methods": "POST,OPTIONS"
  };

  // กรณีเป็น Preflight request (OPTIONS)
  if (event.httpMethod === 'OPTIONS') {
    return { statusCode: 200, headers, body: '' };
  }

  try {
    const body = JSON.parse(event.body || '{}');
    const { studentId, activityId } = body;

    // 1. ตรวจสอบข้อมูลที่ส่งมา
    if (!studentId || !activityId) {
      return {
        statusCode: 400,
        headers,
        body: JSON.stringify({ success: false, message: "Missing studentId or activityId" })
      };
    }

    // 2. (Optional) เช็คก่อนว่าเคยสมัครไปแล้วหรือยัง?
    // ป้องกันการกดซ้ำแล้วข้อมูลทับกัน
    const checkDuplicate = await docClient.send(new QueryCommand({
      TableName: TABLE_NAME,
      // ✅ แก้ไข: รวมเงื่อนไข activityId เข้ามาในนี้เลย
      KeyConditionExpression: "studentId = :sid AND activityId = :aid",
      // ลบ FilterExpression ทิ้งไปเลย
      ExpressionAttributeValues: {
        ":sid": studentId,
        ":aid": activityId
      }
    }));

    if (checkDuplicate.Items && checkDuplicate.Items.length > 0) {
      return {
        statusCode: 400, // หรือ 200 แล้วบอกว่าสมัครไปแล้วก็ได้
        headers,
        body: JSON.stringify({ success: false, message: "คุณได้สมัครกิจกรรมนี้ไปแล้ว" })
      };
    }

    // 3. เตรียมข้อมูลที่จะบันทึก (สร้างตาม Schema ที่คุณให้มา)
    const timestamp = new Date().toISOString();
    const participationId = "part" + randomUUID().split('-')[0]; // สร้าง ID สั้นๆ เช่น part41011ddf

    const newItem = {
      studentId: studentId,                // Partition Key
      activityId: activityId,              // Sort Key (แนะนำให้ตั้งเป็น Sort Key ถ้าตารางออกแบบไว้)
      participationId: participationId,
      
      // สถานะเริ่มต้นของการสมัคร
      isConfirmed: false,                  // เพิ่งสมัคร ยังไม่ได้เข้าร่วมจริง
      confirmedAt: null,                   // ยังไม่มีเวลาเข้าร่วม
      
      surveyCompleted: false,              // ยังไม่ได้ทำแบบประเมิน
      surveyCompletedAt: null,
      
      registeredAt: timestamp,             // เวลาที่กดปุ่มสมัคร
      createdAt: timestamp,
      updatedAt: timestamp
    };

    // 4. บันทึกลง DynamoDB
    await docClient.send(new PutCommand({
      TableName: TABLE_NAME,
      Item: newItem
    }));

    // 5. ส่งค่าตอบกลับสำเร็จ
    return {
      statusCode: 200,
      headers,
      body: JSON.stringify({
        success: true,
        message: "Registration successful",
        data: newItem
      })
    };

  } catch (error) {
    console.error("Error:", error);
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({
        success: false,
        message: "Internal Server Error",
        error: error.message
      })
    };
  }
};